### PR TITLE
Support true owner and group by Alluxio FUSE with allow_other options

### LINKED ISSUE #9552: Can't distinguish creator of file by Alluxio FUSE with allow_other options
**Alluxio Version:**
2.0

**Describe the bug**
When mount Alluxio FUSE by user alluxio, if another user creates a file or directory in the mount point, the owner of the file or directory will be the user alluxio. I can't distinguish who is the true owner.

**To Reproduce**
mount Alluxio FUSE by user alluxio
```
alluxio $ alluxio-fuse mount -o allow_other /mnt/alluxio-fuse/ /
```
mkdir by user hahaha
```
hahaha $ cd /mnt/alluxio-fuse
hahaha $ mkdir hahaha
hahaha $ ls -l -d hahaha
drwxr-xr-x 1 alluxio alluxio 0 Jul 24 15:09 hahaha
```

### PR DESCRIPTION
**Pull request by liuhongtong**
fixes #9552

Signed-off-by: liuhongtong <hongtongliu@126.com>

### CODE REVIEW

**Comment by liuhongtong** [comment]
@apc999 @LuQQiu PTAL.

**Comment by liuhongtong** [comment]
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 604.184 sec <<< FAILURE! - in alluxio.multi.process.MultiProcessClusterTest
zookeeper(alluxio.multi.process.MultiProcessClusterTest)  Time elapsed: 600.009 sec  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 600000 milliseconds
	at alluxio.multi.process.MultiProcessClusterTest.createAndOpenFile(MultiProcessClusterTest.java:76)
	at alluxio.multi.process.MultiProcessClusterTest.zookeeper(MultiProcessClusterTest.java:63)

**Comment by LuQQiu** [comment]
@liuhongtong Thanks for finding the information of how to get the context uid/gid!! Awesome!!!!
I suggest targeting master branch for this new feature since 1.8.2 just code freeze and we are not sure when will be the 1.8.3 or if there will be an 1.8.3. Even there will be 1.8.3, we are likely to only want to fix major bugs in it instead of adding new features.

**Comment by LuQQiu** [comment]
@liuhongtong For the `CreateFileOptions` and `CreateDirectoryOptions` in the client side, we didn't allow setting Owner:Group before. This is because we do not want users to arbitrary setting owner or group, this may introduce security issues. We want them to explicitly setting owner and group via `SetAttribute` call. For 1.8, I suggest we go for smaller changes, use existing `CreateFile/CreateDirectory` APIs and make one more `setAttribute` call.

**Comment by liuhongtong** [comment]
@LuQQiu I will push a PR based on master branch later.

**Comment by liuhongtong** [comment]
@LuQQiu Maybe this PR can be cherry-picked to master branch. PTAL if jenkins works well.

**Comment by liuhongtong** [comment]
@LuQQiu PTAL

**Comment by liuhongtong** [comment]
@LuQQiu I have checked the source code in libfuse. libfuse shares req with jnr-fuse by pthread_setspecific. So we can always be able to get the owner and group.

**Comment by gpang** [comment]
@liuhongtong @LuQQiu This PR should target `master` branch first, and then be cherry-picked to other release branches (2.0, 1.8). Thanks!

**Comment by LuQQiu** [comment]
@liuhongtong Please help create a PR targeting master branch and we will take care of the following cherry-pick to 2.0/1.8 processes.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java [hunk 1]
```diff
@@ -235,6 +251,8 @@ public int create(String path, @mode_t long mode, FuseFileInfo fi) {
         // Assuming I will never wrap around (2^64 open files are quite a lot anyway)
         mNextOpenFileId += 1;
       }
+      mFileSystem.setAttribute(uri,
```
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: Can we store the following fields for FuseFileSystem to get the default user/group running this process:
```
    mFuseOwner = SecurityUtils.getOwnerFromLoginModule();
    mFuseGroup = SecurityUtils.getGroupFromLoginModule();
```
Then compare compare `userName` with `mFuseOwner` and `groupName` with `mFuseGroup` and only makes the RPC on string difference? In this way, we could avoid making extra RPCs in users/groups are the same as the default

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java [hunk 2]
```diff
@@ -389,11 +407,29 @@ public String getFSName() {
...
+      }
+      String userName = AlluxioFuseUtils.getUserName(uid);
+      if (userName.isEmpty()) {
+        // This should never be reached since input uid is always valid
+        LOG.error("Failed to get user name from uid {}", uid);
+        return -ErrorCodes.EFAULT();
+      }
+      mFileSystem.createDirectory(turi, CreateDirectoryOptions.defaults()
+          .setMode(new Mode((short) mode)));
+      mFileSystem.setAttribute(turi, SetAttributeOptions.defaults()
```
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: same comment. only run the `setAttribute` RPC on different user or group name?

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java (hunk 1 above)
Comment (apc999) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: To derive similar values in master or 2.0, check https://github.com/Alluxio/alluxio/blob/1d50c373dcfa44ec4fa1b8bd05ab5cd49e72bd97/core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java#L133

**Comment by apc999** [comment]
ping @liuhongtong

**Comment by liuhongtong** [comment]
@apc999 maybe failed test doesn't matter to me

Tests run: 9, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 20.288 sec <<< FAILURE! - in alluxio.client.cli.fs.command.CopyToLocalCommandIntegrationTest
copyToLocalLarge(alluxio.client.cli.fs.command.CopyToLocalCommandIntegrationTest)  Time elapsed: 2.134 sec  <<< FAILURE!
org.junit.ComparisonFailure: 
expected:<C[opied /testFile to file:///tmp/alluxio-tests/test-cluster-6174caa0-3c11-4ddc-9ca9-0ef8f55108a8/testFile]
> but was:<C[hannel to b3c3a7d4a35f/172.17.0.4:32818: blockId 16777216 not found]
>
	at alluxio.client.cli.fs.command.CopyToLocalCommandIntegrationTest.copyToLocalLarge(CopyToLocalCommandIntegrationTest.java:65)

**Comment by apc999** [comment]
Jenkins, test this please.

**Comment by apc999** [comment]
@LuQQiu since this PR is targeting 1.8, can you help create the same patch based on @liuhongtong 's change here targeting master branch first, then let's cherry-pick the patch to 2.0 and 1.8 branch

**Comment by LuQQiu** [comment]
@apc999 No problem

**Comment by apc999** [comment]
taken over by https://github.com/Alluxio/alluxio/pull/9950 which is a real cherry-pick from master branch

**Pull request closed without merging**